File: cpp/acwing/high_precision_mul.cpp
//
// Created by bruce on 2021/6/18.
//


/**
 * AcWing-793. 高精度乘法
 * 题目链接：https://www.acwing.com/problem/content/795/
 * A x b = C -> 将 b 看成一个整体，A 一个一个取值
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * C = A * b
 * @param A
 * @param b
 * @return
 */
vector<int> mul(vector<int> &A, int b) {
    vector<int> C;

    int t = 0; // 进位
    for (int i = 0; i < A.size(); i++) {
        t = A[i] * b + t;
        C.push_back(t % 10);
        t /= 10;
    }

    while (t != 0) {
        C.push_back(t % 10);
        t /= 10;
    }

    while (C.size() > 1 && C.back() == 0) C.pop_back();

    return C;
}

int main() {
    string a;
    int b;
    vector<int> A, B;

    cin >> a >> b;
    for (int i = a.size() - 1; i >= 0; i--) A.push_back(a[i] - '0');

    auto C = mul(A, b);

    for (int i = C.size() - 1; i >= 0; i--) printf("%d", C[i]);

    return 0;
}

File: js/acwing/high_precision_mul.js
/**
 * @Description AcWing-793. 高精度乘法
 * 题目链接：https://www.acwing.com/problem/content/795/
 * @Author PAN Bruce
 * @Date 2021/6/18
 */

var fs = require('fs');
var buf = '';

process.stdin.on("readable", function () {
    var chunk = process.stdin.read();
    if (chunk) buf += chunk.toString();
});

process.stdin.on("end", function () {
    const lines = buf.split("\n");
    // 输入：共两行，每行包含一个整数。
    let A = lines[0].trim().split("").reverse();
    let b = +lines[1].trim();
    let C = mul(A, b);
    // 输出：共一行
    console.log(C.reverse().join(""));
});

/**
 * C = A * b
 * @param A
 * @param b
 */
var mul = function (A, b) {
    let C = [];

    for (let i = 0, temp = 0; i < A.length || temp; i++) {
        if (i < A.length) temp = b * A[i] + temp;
        C.push(temp % 10);
        temp = Math.floor(temp / 10);
    }

    // 处理前导0
    while (C.length > 1 && C[C.length - 1] === 0) C.pop();

    return C;
};
